src/providers/Wallet.js

```diff
@@ -13,23 +20,122 @@ function useWalletAugmented() {
 
 // Adds Ethers.js to the useWallet() object
 function WalletAugmented({ children }) {
+  // This new state is neede to avoid showing the user the profile signature request on every network change
+  const [chainReseted, setChainReseted] = useState(false)
+  const [chainId, setChainId] = useState(-1)
+  /* We need  to pass down on the providers tree a preferred network in case that there is no network connnected
+  or the connected network is not supported in order to show some data and also to react to the network drop down selector changes */
+  const [preferredNetwork, setPreferredNetwork] = useState(getPreferredChain())
+
   const wallet = useWallet()
   const { ethereum } = wallet
 
   const ethers = useMemo(() => {
     if (!ethereum) {
-      return getDefaultProvider()
+      return getPreferredProvider()
     }
 
-    const ensRegistry = getNetwork()?.ensRegistry
+    const ensRegistry = getNetwork(wallet.chainId)?.ensRegistry
     return new EthersProviders.Web3Provider(ethereum, {
       name: '',
-      chainId: getDefaultChain(),
+      chainId: preferredNetwork,
       ensAddress: ensRegistry,
     })
-  }, [ethereum])
+  }, [ethereum, preferredNetwork]) //eslint-disable-line
+
+  const isSupportedNetwork = useMemo(() => {
+    return isSupportedChain(wallet.chainId)
+  }, [wallet.chainId])
+
+  const connect = useCallback(async () => {
+    const connectedAddresses = await window.ethereum.request({
+      method: 'eth_accounts',
+    })
+    if (connectedAddresses.length > 0) {
+      try {
+        await wallet.connect('injected')
+        console.log('WALLET ', wallet)
+      } catch (e) {
+        console.error(e)
+      }
+    }
+  }, [wallet])
+
+  const resetConnection = useCallback(async () => {
+    setChainReseted(true)
+    setChainId(-1)
+    await wallet.reset()
+  }, [wallet])
+
+  // Handle connect automatically if windows.ethereum is available and we have some connected address on the wallet
+  useEffect(() => {
+    async function connectNetwork() {
+      await connect()
+    }
+    if (window.ethereum && wallet.chainId === -1) {
+      connectNetwork()
+    }
+    if (isSupportedNetwork) {
+      setChainId(wallet.chainId)
+      setPreferredNetwork(wallet.chainId)
+    }
+  }, []) //eslint-disable-line
+
+  useEffect(() => {
+    async function connectNetwork() {
+      await resetConnection()
+      await connect()
+    }
+
+    if (
+      wallet.chainId !== wallet._web3ReactContext.chainId &&
+      wallet.status !== 'connecting'
+    ) {
+      if (isSupportedChain(wallet._web3ReactContext.chainId)) {
+        connectNetwork()
+      } else {
+        resetConnection()
+      }
+    }
+    if (isSupportedChain(wallet._web3ReactContext.chainId)) {
+      setChainId(wallet.chainId)
+      setPreferredNetwork(wallet.chainId)
+    }
```

Comment:
Would you mind give me a bit of context on what we are doing here? I am not fully understanding what's happening.